src/Symfony/Component/DomCrawler/DomCrawler.php

```diff
@@ -0,0 +1,350 @@
+        }
+
+        $internalErrors = libxml_use_internal_errors(true);
+
+        if ($this->options & self::CRAWLER_USE_NATIVE_PARSER) {
+            try {
+                $dom = \DOM\XMLDocument::createFromString($content, $options);
+            } catch (\Exception) {
+                // like with legacy nodes, create an empty document if
+                // content cannot be loaded
+                $dom = \DOM\XMLDocument::createEmpty();
+            }
+        } else {
+            $dom = new \DOMDocument('1.0', $charset);
+            $dom->validateOnParse = true;
+
+            if ('' !== trim($content)) {
+                @$dom->loadXML($content, $options);
+            }
+        }
+
+        libxml_use_internal_errors($internalErrors);
+
+        $this->addDocument($dom);
+
+        $this->isHtml = false;
+    }
+
+    /**
+     * Adds a \DOMDocument or a \DOM\Document to the list of nodes.
+     */
+    public function addDocument(\DOMDocument|\DOM\Document $dom): void
+    {
+        if ($dom->documentElement) {
+            $this->addNode($dom->documentElement);
+        }
+    }
+
+    /**
+     * Adds a \DOMNodeList or a \DOM\NodeList to the list of nodes.
+     */
+    public function addNodeList(\DOMNodeList|\DOM\NodeList $nodes): void
+    {
+        foreach ($nodes as $node) {
+            if ($node instanceof \DOMNode || $node instanceof \DOM\Node) {
+                $this->addNode($node);
+            }
+        }
+    }
+
+    /**
+     * Adds a \DOMNode or a \DOM\Node instance to the list of nodes.
+     */
+    public function addNode(\DOMNode|\DOM\Node $node): void
+    {
+        if ($node instanceof \DOMDocument || $node instanceof \DOM\Document) {
+            $node = $node->documentElement;
+        }
+
+        if (null !== $this->domDocument && $this->domDocument !== $node->ownerDocument) {
+            throw new \InvalidArgumentException('Attaching DOM nodes from multiple documents in the same crawler is forbidden.');
+        }
+
+        $this->domDocument ??= $node->ownerDocument;
+        if ($node->ownerDocument instanceof \DOMDocument) {
+            $this->document = $node->ownerDocument;
+        }
+
+        // Don't add duplicate nodes in the Crawler
+        if (\in_array($node, $this->nodes, true)) {
+            return;
+        }
+
+        $this->nodes[] = $node;
+    }
+
+    protected function sibling(\DOMNode|\DOM\Node $node, string $siblingDir = 'nextSibling'): array
+    {
+        $nodes = [];
+
+        $currentNode = $this->getDomNode(0);
+        do {
+            if ($node !== $currentNode && \XML_ELEMENT_NODE === $node->nodeType) {
+                $nodes[] = $node;
+            }
+        } while ($node = $node->$siblingDir);
+
+        return $nodes;
+    }
+
+    private function parseHtml5(string $htmlContent, string $charset = 'UTF-8'): \DOMDocument|\DOM\HTMLDocument
+    {
+        $htmlContent = $this->convertToHtmlEntities($htmlContent, $charset);
```

Comment:
is this actually needed when using the native parser ? Or does it handle UTF-8 properly ?

Reply:
It is needed. If I remove this call, I gave the following failure:

```
There was 1 failure:

1) Symfony\Component\DomCrawler\Tests\DomCrawlerNativeParserTest::testAddHtmlContentCharsetGbk
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'中文'
+'����'
```

Reply:
FWIW, it works if you pass in the character set explicitly, it's just the auto detection that falls back to UTF-8 in this case because it can't detect the encoding through mime / meta-tag sniffing.
e.g.
```php
<?php

$html = base64_decode('PGh0bWw+PHA+1tDOxDwvcD48L2h0bWw+');
$doc = DOM\HTMLDocument::createFromString($html, 0, "GBK");
var_dump($doc->firstElementChild->firstElementChild->nextElementSibling->textContent);
```

Yields the expected result.

Reply:
Works like a charm. Thanks!